orttraining/orttraining/python/training/onnxblock/_graph_utils.py

```diff
@@ -0,0 +1,205 @@
+    graph_outputs = []
+
+    lazy_reset_grad_input_name = "lazy_reset_grad"
+    gradient_accumulation_name = "accumulation"
+    gradient_buffer_name_suffix = "buffer"
+    gradient_output_name_suffix = "out"
+
+    for idx, graph_output in enumerate(grad_model.graph.output):
+        if graph_output.name not in gradient_output_names:
+            # If graph output is not a gradient, there is no
+            # need to build the gradient accumulator node for it.
+            graph_outputs.append(graph_output)
+            continue
+
+        # gradient accumulation node inputs and output names
+        grad_name = graph_output.name
+        grad_accumulation_buffer_name = (
+            f"{grad_name}.{gradient_accumulation_name}.{gradient_buffer_name_suffix}"
+        )
+        grad_accumulation_output_name = (
+            f"{grad_name}.{gradient_accumulation_name}.{gradient_output_name_suffix}"
+        )
+
+        # Gradient accumulation node
+        acc_node = onnx.helper.make_node(
+            "InPlaceAccumulator",
+            [grad_accumulation_buffer_name, grad_name, lazy_reset_grad_input_name],
+            [grad_accumulation_output_name],
+            name=f"GradientAccumulator{idx}",
+            domain="com.microsoft",
+        )
+
+        graph_nodes.append(acc_node)
+
+        # grad buffer is also a graph input
+        grad_accumulation_buffer_input = copy.deepcopy(graph_output)
+        grad_accumulation_buffer_input.name = grad_accumulation_buffer_name
+        graph_inputs.append(grad_accumulation_buffer_input)
+
+        # accumulated gradient is also a graph output
+        grad_accumulation_output = copy.deepcopy(graph_output)
+        grad_accumulation_output.name = grad_accumulation_output_name
+        graph_outputs.append(grad_accumulation_output)
+
+    lazy_reset_grad_input = onnx.helper.make_tensor_value_info(
+        lazy_reset_grad_input_name, onnx.TensorProto.BOOL, [1]
+    )
+    graph_inputs.append(lazy_reset_grad_input)
+
+    del grad_model.graph.output[:]
+    grad_model.graph.output.extend(graph_outputs)
+
+
+def get_model_parameters(model, args_not_requiring_gradient):
+    """Returns trainable and non trainable onnx model parameters."""
+
+    trainable_params = []
+    non_trainable_params = []
+    for initializer in model.graph.initializer:
+        # All model parameters should have their names not begin with
+        # a digit. So, check to see if the initializer's first char
+        # is a digit. If not, it is either a trainable, or a non
+        # trainable parameter.
+        # Note that this assumption can be made because the export logic
+        # does not change the names of the original model parameters.
+        # and the original model parameters don't have their names begin
+        # with a digit.
+        # On the other hand, const initializers are generated by export
+        # logic and have a digit prefix.
+        # TODO: validate this assumption. If assumption is not valid,
```

Comment:
Yeah, maybe we can just return all non-input initializer as parameters?

Reply:
Then by default, all param is non-trainable, the one set explicitly by SetRequiresGrad will be trainable.